EVPN VxLAN enhancement to support P2MP tunnel based programming for Layer2 extension

dgsudharsan opened the pull request:
<!--
Please make sure you have read and understood the contribution guildlines:
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

1. Make sure your commit includes a signature generted with `git commit -s`
2. Make sure your commit title follows the correct format: [component]: description
3. Make sure your commit message contains enough details about the change and related tests
4. Make sure your pull request adds related reviewers, asignees, labels

Please also provide the following information in this pull request:
-->

**What I did**
Added support to identify based on sai attribute capability if P2P tunnels are support and if not achieve L2 extension using P2MP tunnels. This involves adding P2MP tunnel bridgeport, remote endpoint IP combination to L2MC group in VLAN that is using combined mode.

**Why I did it**
Certain platforms support P2P tunnels and while few prefer using P2MP tunnels for L2 extension due to resource constraints.

**How I verified it**
Added VS test cases to cover all the existing scenarios using P2MP model for programming SAI.

**Details if related**

dgsudharsan commented on orchagent/vxlanorch.cpp:
```diff
@@ -1058,6 +1062,43 @@ void VxlanTunnel::updateDipTunnelRefCnt(bool inc, tunnel_refcnt_t& tnl_refcnts,
+        }
+        else
+        {
+            it->second.ip_refcnt++;
```
Both are same.  auto it = tnl_users_.find(remote_vtep);  -->If remote vtep exists in the map it.second will be referencing the value of tnl_users_[remote_vtep]

dgsudharsan commented on orchagent/vxlanorch.cpp:
```diff
@@ -1058,6 +1062,43 @@ void VxlanTunnel::updateDipTunnelRefCnt(bool inc, tunnel_refcnt_t& tnl_refcnts,
     }
 }
 
+void VxlanTunnel::updateRemoteEndPointIpRef(const std::string remote_vtep, bool inc)
```
The IMR reference count is incremented in createDIP tunnel flow. However in case of P2MP the reference for the remote can be inferred on the bridge ports added to L2MC group member. So it is not separately maintained. Please check isTunnelReferenced
This not only maintains the remote end point ip but also remote VNI too.
        /* Bridge port will have reference since on IMET routes reception L2MC group member
           would be created with end point IP and the P2MP tunnel bridge port */

        if (gPortsOrch->getBridgePortReferenceCount(tunnelPort) != 0)

dgsudharsan replied:
Routes flow come through route orch and thus need to be maintained. So I am reusing existing route reference while considering bridge port references for remote IMR, VNI

dgsudharsan commented on orchagent/vxlanorch.h:
```diff
@@ -338,6 +354,10 @@ class VxlanTunnelOrch : public Orch2
         vxlan_vni_vlan_map_table_.erase(vni);
     }
 
+    bool dipTunnelsUsed(void)
```
Done

dgsudharsan commented on orchagent/vxlanorch.h:
```diff
@@ -349,6 +369,7 @@ class VxlanTunnelOrch : public Orch2
     VxlanVniVlanMapTable vxlan_vni_vlan_map_table_;
     VTEPTable vtep_table_;
     Table m_stateVxlanTable;
+    bool dip_tunnels_used;
```
Done

dgsudharsan commented on orchagent/vxlanorch.h:
```diff
@@ -35,6 +35,15 @@ typedef enum
+#define MIN_VLAN_ID 1
+#define MAX_VLAN_ID 4095
+
+#define MAX_VNI_ID (1<<24)
```
Done

dgsudharsan commented on orchagent/vxlanorch.cpp:
```diff
@@ -1439,6 +1480,12 @@ bool  VxlanTunnelOrch::addTunnelUser(const std::string remote_vtep, uint32_t vni
         return false;
     }
 
+    if (!dipTunnelsUsed() && usr == TUNNEL_USER_IP)
```
Makes sense. Will remove the user check.

dgsudharsan commented on orchagent/vxlanorch.cpp:
```diff
@@ -1479,12 +1526,29 @@ bool  VxlanTunnelOrch::delTunnelUser(const std::string remote_vtep, uint32_t vni
 
     Port tunnelPort;
     gPortsOrch->getPort(port_tunnel_name,tunnelPort);
```
That's a good catch. Updated the changes to use P2MP tunnel port in the scenario

dgsudharsan commented on orchagent/vxlanorch.cpp:
```diff
@@ -1520,14 +1584,30 @@ void VxlanTunnelOrch::deleteTunnelPort(Port &tunnelPort)
         return;
     }
 
+    if (isSrcVtepTunnel(tunnelPort)) 
```
Agreed. I removed this and replaced with isDipTunnelsSupported

dgsudharsan commented on orchagent/vxlanorch.cpp:
```diff
@@ -1657,6 +1753,49 @@ bool VxlanTunnelOrch::getTunnelPort(const std::string& remote_vtep,Port& tunnelP
+        return false;
+    }
+
+    if (tunnelPort.m_fdb_count != 0)
```
Moved it inside P2MP logic. I felt the same reasoning is applicable for P2P bridge port too but agree it's better not to change existing flows.

dgsudharsan commented on orchagent/vxlanorch.cpp:
```diff
@@ -1657,6 +1753,49 @@ bool VxlanTunnelOrch::getTunnelPort(const std::string& remote_vtep,Port& tunnelP
+        return true;
+    }
+
+    if (dip_tunnels_used)
```
Done

dgsudharsan commented:
@srj102 I have given explanation for some review comments and the rest I have addressed in code (marked those as resolved).

srj102 commented on orchagent/vxlanorch.h:
```diff
@@ -190,11 +203,13 @@ class VxlanTunnel
     int getDipTunnelRefCnt(const std::string);
     int getDipTunnelIMRRefCnt(const std::string);
-    int getDipTunnelIPRefCnt(const std::string);
+    int getRemoteEndPointIPRefCnt(const std::string);
```
Sure I understand that we would like a naming common for P2P and P2MP.  So it might be better to change the naming for the IMR as well as the total refcnt i.e. we can eliminate the use of DipTunnel in the name for all the common cases.

dgsudharsan replied:
Modified as suggested.

srj102 commented on orchagent/fdborch.cpp:
```diff
@@ -1147,7 +1161,7 @@ bool FdbOrch::addFdbEntry(const FdbEntry& entry, const string& port_name,
 
     /* Retry until port is member of vlan*/
-    if (vlan.m_members.find(port_name) == vlan.m_members.end())
+    if (check_vlan_member && vlan.m_members.find(port_name) == vlan.m_members.end())
```
I see that the EvpnRemoteVnip2mpOrch::addOperation already has a call to 
gPortsOrch->addVlanMember in which case I don't see why this check is required. 
if we call portsorch->isVlanMember.. It should work for both P2P and P2MP ?

dgsudharsan replied:
Actually I thought of a better logic to include checks for P2P and P2MP. In P2MP scenario, the vlan member differentiator is remote IP (since only one port exists based on SIP tunnel) and thus including that in check will  ensure it works in both scenarios.

srj102 commented on orchagent/fdborch.cpp:
```diff
@@ -1121,15 +1135,15 @@ void FdbOrch::updateVlanMember(const VlanMemberUpdate& update)
 
 bool FdbOrch::addFdbEntry(const FdbEntry& entry, const string& port_name,
-        FdbData fdbData)
+        FdbData fdbData, bool check_vlan_member)
```
Do we need to change the API signature ? If there is a need to check for vlan member it can be done using the isDipTunnelSupported call ?

dgsudharsan replied:
Removed changes to API signature.

dgsudharsan commented on orchagent/fdborch.cpp:
```diff
@@ -804,9 +821,6 @@ void FdbOrch::doTask(Consumer& consumer)
                     port = tunnel_orch->getTunnelPortName(remote_ip);
                 }
-
-
```
That's a good catch. It got removed during a merge conflict. I have added it back.

dgsudharsan commented:
@srj102 Can you please review? I have addressed your review comments.

srj102 commented:
> @srj102 Can you please review? I have addressed your review comments.

Thanks for taking care of the comments. Changes look good.

dgsudharsan commented:
> > @srj102 Can you please review? I have addressed your review comments.
> 
> Thanks for taking care of the comments. Changes look good.

Thank you Rajesh. Can you please approve?

dgsudharsan commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

dgsudharsan commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

dgsudharsan commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

dgsudharsan commented:
/azpw run

mssonicbld commented:
/AzurePipelines run
